### PR TITLE
[all][server][dvc] Fast-avro upgrade, fix meta system store schema update issue and partition size race condition

### PR DESCRIPTION
**Pull request by gaojieliu**
This code change includes several fixes:
1. Bump up fast-avro lib tot pick up the recent serde optimizations.
2. Fix the meta system store schema race condition issue in Venice Server. Previously, the update schema may not be available for the latest superset schema since there is a delay between registering superset schema and its derived schema. The fix will always retrieve a latest value schema with a valid derived schema.
3. Use RocksDB API to fetch partition size to avoid the following race condition: RocksDB log compaction can delete file in the middle of measuring the directory size.
4. Let 'fast.avro.field.limit.per.method' to control fast serializer size as well.

## How was this PR tested?
CI

## Does this PR introduce any user-facing changes?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure to explain your proposed changes and call out the behavior change.

### CODE REVIEW

File: internal/venice-common/src/main/java/com/linkedin/venice/meta/ReadOnlySchemaRepository.java [hunk 1]
```diff
@@ -68,4 +69,29 @@ public interface ReadOnlySchemaRepository extends VeniceResource {
...
+      /**
+       * Can't find the derived schema for the latest value schema, so it will fall back to find out the latest
+       * value schema, which has a valid derived schema.
+       */
+      Collection<DerivedSchemaEntry> derivedSchemaEntries = getDerivedSchemas(storeName);
+      DerivedSchemaEntry latestDerivedSchemaEntry = null;
+      for (DerivedSchemaEntry entry: derivedSchemaEntries) {
+        if (latestDerivedSchemaEntry == null || entry.getValueSchemaID() > latestDerivedSchemaEntry.getValueSchemaID()
+            || entry.getValueSchemaID() == latestDerivedSchemaEntry.getValueSchemaID()
+                && entry.getId() > latestDerivedSchemaEntry.getId()) {
```
Comment (FelixGV) [inline internal/venice-common/src/main/java/com/linkedin/venice/meta/ReadOnlySchemaRepository.java:87]: Minor nitpick on readability... I find it a bit hard to parse a sequence of mixed boolean operators... IIUC, `&&` has higher precedence than `||`, so the if condition `A || B || C && D` is equivalent to `A || B || (C && D)`. Although they are equivalent, I find the latter easier to parse visually. Or alternatively, the whole expression might be more concise and clear if using an extra function, e.g. `A || B || func(C, D)`, where `func(C, D) { return C && D; }`. What do you think?

File: services/venice-server/src/main/java/com/linkedin/avro/fastserde/FastDeserializerGeneratorAccessor.java [hunk 2]
```diff
@@ -6,5 +6,6 @@
 public class FastDeserializerGeneratorAccessor {
   public static void setFieldsPerPopulationMethod(int limit) {
     FastDeserializerGenerator.setFieldsPerPopulationMethod(limit);
+    FastSerializerGenerator.setFieldsPerRecordSerializationMethod(limit);
```
Comment (FelixGV) [inline services/venice-server/src/main/java/com/linkedin/avro/fastserde/FastDeserializerGeneratorAccessor.java:9]: We should also mention this in the commit message.

File: internal/venice-common/src/main/java/com/linkedin/venice/meta/ReadOnlySchemaRepository.java [hunk 3]
```diff
@@ -68,4 +69,29 @@ public interface ReadOnlySchemaRepository extends VeniceResource {
...
+  default DerivedSchemaEntry getLatestDerivedSchema(String storeName) {
+    SchemaEntry valueSchemaEntry = getSupersetOrLatestValueSchema(storeName);
+    try {
+      return getLatestDerivedSchema(storeName, valueSchemaEntry.getId());
+    } catch (Exception e) {
+      /**
+       * Can't find the derived schema for the latest value schema, so it will fall back to find out the latest
+       * value schema, which has a valid derived schema.
+       */
+      Collection<DerivedSchemaEntry> derivedSchemaEntries = getDerivedSchemas(storeName);
```
Comment (sixpluszero) [inline internal/venice-common/src/main/java/com/linkedin/venice/meta/ReadOnlySchemaRepository.java:82]: I am a bit confused about this, why don't we do below logic directly? Given that it should always return the known latest one.

Comment (gaojieliu) [reply internal/venice-common/src/main/java/com/linkedin/venice/meta/ReadOnlySchemaRepository.java:82]: The latest superset schema might not be the one with the max schema id.

**Review by FelixGV — COMMENTED** [review]
Thanks! LGTM, I just have some minor comments...

File: internal/venice-common/src/main/java/com/linkedin/venice/meta/ReadOnlySchemaRepository.java (hunk 3 above)
Comment (sixpluszero) [reply internal/venice-common/src/main/java/com/linkedin/venice/meta/ReadOnlySchemaRepository.java:82]: Oh...right. This has to be the fallback logic. 

But just wondering if we evolve schema with field reduction, latest value schema and update schema is still not correct right? Even if we use the fall back logic to find a latest derived schema, it might not work? In this case we should still bounce the server to mitigate but SIT will fail still I believe?

Comment (gaojieliu) [reply internal/venice-common/src/main/java/com/linkedin/venice/meta/ReadOnlySchemaRepository.java:82]: No, `SIT` will use the returned `DerivedSchemaEntry` to extract the corresponding value schema.

File: services/venice-server/src/main/java/com/linkedin/avro/fastserde/FastDeserializerGeneratorAccessor.java (hunk 2 above)
Comment (gaojieliu) [reply services/venice-server/src/main/java/com/linkedin/avro/fastserde/FastDeserializerGeneratorAccessor.java:9]: Sure.

File: internal/venice-common/src/main/java/com/linkedin/venice/meta/ReadOnlySchemaRepository.java (hunk 1 above)
Comment (gaojieliu) [reply internal/venice-common/src/main/java/com/linkedin/venice/meta/ReadOnlySchemaRepository.java:87]: Sure, I will tweak the if condition.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/store/rocksdb/RocksDBStoragePartition.java [hunk 4]
```diff
@@ -923,20 +922,14 @@ public boolean verifyConfig(StoragePartitionConfig partitionConfig) {
         && writeOnly == partitionConfig.isWriteOnlyConfig();
   }
 
-  /**
-   * This method calculates the file size by adding all subdirectories size
-   * @return the partition db size in bytes
-   */
   @Override
   public long getPartitionSizeInBytes() {
```
[the remark this answers is not in the document]
Comment (FelixGV) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/store/rocksdb/RocksDBStoragePartition.java:926]: There are lots of `synchronized` functions in `AbstractStorageEngine`, but none of them look to me like hot path functions...

[the remark this answers is not in the document]
Comment (sushantmane) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/store/rocksdb/RocksDBStoragePartition.java:926]: Oh ok.

[the remark this answers is not in the document]
Comment (gaojieliu) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/store/rocksdb/RocksDBStoragePartition.java:926]: Yeah, also all the writes to the same partitions are sequential, so the perf impact should be minimal.
